go/chat/teams.go

```diff
@@ -263,6 +263,45 @@ func (t *TeamsNameInfoSource) EphemeralDecryptionKey(ctx context.Context, tlfNam
 	return t.G().GetEKLib().GetTeamEK(ctx, teamID, generation)
 }
 
+func (t *TeamsNameInfoSource) ShouldPairwiseMAC(ctx context.Context, tlfName string, tlfID chat1.TLFID,
+	membersType chat1.ConversationMembersType, public bool) (bool, []keybase1.KID, error) {
+	return shouldPairwiseMAC(ctx, t.G(), t.loader, tlfName, tlfID, membersType, public)
+}
+
+func shouldPairwiseMAC(ctx context.Context, g *globals.Context, loader *TeamLoader, tlfName string,
+	tlfID chat1.TLFID, membersType chat1.ConversationMembersType, public bool) (bool, []keybase1.KID, error) {
+	team, err := loader.loadTeam(ctx, tlfID, tlfName, membersType, public, nil)
+	if err != nil {
+		return false, nil, err
+	}
+	members, err := team.Members()
+	if err != nil {
+		return false, nil, err
+	}
+	memberUVs := members.AllUserVersions()
+
+	// For performance reasons, we don't try to pairwise MAC any messages in
+	// large teams.
+	if len(memberUVs) > libkb.MaxTeamMembersForPairwiseMAC {
+		return false, nil, nil
+	}
+
+	unrevokedKIDs := []keybase1.KID{}
+	for _, member := range memberUVs {
+		upak, _, err := g.GetUPAKLoader().LoadV2(libkb.NewLoadUserByUIDArg(ctx, g.GlobalContext, member.Uid))
+		if err != nil {
+			return false, nil, err
+		}
+		for _, key := range upak.Current.DeviceKeys {
+			// Include only unrevoked encryption keys.
+			if !key.Base.IsSibkey && key.Base.Revocation == nil {
+				unrevokedKIDs = append(unrevokedKIDs, key.Base.Kid)
```

Comment:
I'd prefer to avoid putting a silent downgrade in here. (E.g. Charlie could "maliciously provision" a crap ton of devices, which makes Alice sign messages she thought she wasn't signing.)

Reply:
But I'll leave a comment for future generations.